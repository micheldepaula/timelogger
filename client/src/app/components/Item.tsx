import { useAppDispatch } from '../../app/hooks'
import React, { useState } from 'react'
import { IProject } from '../../utils'
import { editProject } from '../redux/projectSlice'

type ItemProps = {
  project: IProject
}

export default function Item(props: ItemProps) {
  const [edit, setEdit] = useState<boolean>(false)
  const [hours, setHours] = useState(props.project.hours)
  const [name, setName] = useState(props.project.name)
  const dispatch = useAppDispatch()
  const handleEditProject = () => {
    setEdit(false)
    if (props.project.end) {
      dispatch(
        editProject({
          id: props.project.id,
          name: name,
          hours: Number(hours),
          end: new Date(props.project.end.toString()),
        }),
      )
    }
  }
  const EditableInputName = () => {
    return (
      <td className='border px-4 py-2'>
        <input
          className='border rounded-full py-2 px-4'
          value={name}
          type='name'
          placeholder='Project Name'
          aria-label='Project Name'
          onChange={(e) => setName(e.target.value.toString())}
        />
      </td>
    )
  }

  const EditableInputHours = () => {
    return (
      <td className='border px-4 py-2 '>
        <input
          className='border rounded-full py-2 px-4'
          value={hours}
          type='number'
          placeholder='DeadLine'
          aria-label='DeadLine'
          onChange={(e) => setHours(e.target.value.toString())}
        />
      </td>
    )
  }

  const EditButtons = () => {
    return (
      <div className='flex justify-between '>
        <button
          className='bg-blue-500 hover:bg-blue-700 text-white rounded-full py-2 px-4 ml-2'
          onClick={handleEditProject}
        >
          Save
        </button>
        <button
          className='bg-red-500 hover:bg-red-700 text-white rounded-full py-2 px-4 ml-2'
          onClick={() => setEdit(false)}
        >
          cancel
        </button>
      </div>
    )
  }

  return (
    <tr>
      <td className='border px-4 py-2 w-12'>{props.project.id}</td>
      {!edit ? <td className='border px-4 py-2'>{name}</td> : <EditableInputName />}
      {!edit ? <td className='border px-4 py-2'>{hours}</td> : <EditableInputHours />}
      <td className='flex justify-between border px-4 py-2 '>
        {props.project.end && new Date(props.project.end).toISOString().split('T')[0]}
        {!edit ? (
          <button
            className='bg-blue-500 hover:bg-blue-700 text-white rounded-full py-2 px-4 ml-2'
            onClick={() => setEdit(true)}
          >
            Edit
          </button>
        ) : (
          <EditButtons />
        )}
      </td>
    </tr>
  )
}
