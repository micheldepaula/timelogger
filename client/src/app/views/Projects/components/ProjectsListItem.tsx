import { useAppDispatch } from '../../../hooks'
import React, { useState } from 'react'
import { IProject } from '../../../../utils'
import { editProject } from '../../../slice/projectSlice'

type ProjectsListItemProps = {
  project: IProject
}

const ProjectsListItem: React.FC<ProjectsListItemProps> = ({ project }) => {
  const [edit, setEdit] = useState<boolean>(false)
  const [hours, setHours] = useState(project.hours)
  const [name, setName] = useState(project.name)
  const dispatch = useAppDispatch()
  const handleEditProject = () => {
    setEdit(false)
    if (project.end) {
      dispatch(
        editProject({
          id: project.id,
          name: name,
          hours: Number(hours),
          end: new Date(project.end.toString()),
        }),
      )
    }
  }
  const EditableInputName = () => {
    return (
      <td className='border px-4 py-2'>
        <input
          className='border rounded-full py-2 px-4'
          value={name}
          type='name'
          placeholder='Project Name'
          aria-label='Project Name'
          onChange={(e) => setName(e.target.value.toString())}
        />
      </td>
    )
  }

  const EditableInputHours = () => {
    return (
      <td className='border px-4 py-2 '>
        <input
          className='border rounded-full py-2 px-4'
          value={hours}
          type='number'
          placeholder='DeadLine'
          aria-label='DeadLine'
          onChange={(e) => setHours(e.target.value.toString())}
        />
      </td>
    )
  }

  const EditButtons = () => {
    return (
      <div className='flex justify-between '>
        <button
          className='bg-blue-500 hover:bg-blue-700 text-white rounded-full py-2 px-4 ml-2'
          onClick={handleEditProject}
          data-testid={`projects-list-item-${project.id}-button-edit`}
        >
          Save
        </button>
        <button
          className='bg-red-500 hover:bg-red-700 text-white rounded-full py-2 px-4 ml-2'
          onClick={() => setEdit(false)}
          data-testid={`projects-list-item-${project.id}-button-cancel-edit`}
        >
          cancel
        </button>
      </div>
    )
  }

  return (
    <tr data-testid={`projects-list-item-${project.id}`}>
      <td className='border px-4 py-2 w-12'>{project.id}</td>
      {!edit ? <td className='border px-4 py-2'>{name}</td> : <EditableInputName />}
      {!edit ? <td className='border px-4 py-2'>{hours}</td> : <EditableInputHours />}
      <td className='flex justify-between border px-4 py-2 '>
        {project.end && new Date(project.end).toISOString().split('T')[0]}
        {!edit ? (
          <button
            className='bg-blue-500 hover:bg-blue-700 text-white rounded-full py-2 px-4 ml-2'
            onClick={() => setEdit(true)}
          >
            Edit
          </button>
        ) : (
          <EditButtons />
        )}
      </td>
    </tr>
  )
}
export default ProjectsListItem
