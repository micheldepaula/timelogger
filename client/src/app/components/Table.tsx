import React, { useEffect, useState } from 'react'
import Item from './Item'

import { useAppDispatch, useAppSelector } from '../../app/hooks'
import { sortedProjectsAsc, projectsSelector, getProjects } from './../redux/projectSlice'

export default function Table() {
  const [up, setUp] = useState<boolean>(true)
  const dispatch = useAppDispatch()

  const globalData = useAppSelector(projectsSelector)

  useEffect(() => {
    dispatch(getProjects())
  }, [])

  const orderProjectsByEndDate = () => {
    if (up) {
      dispatch(sortedProjectsAsc(globalData.projects))
      setUp(!up)
    } else {
      dispatch({ type: 'projects/sortedProjectsDesc' })
      setUp(!up)
    }
  }

  const IconUp = () => {
    return (
      <svg
        xmlns='http://www.w3.org/2000/svg'
        fill='none'
        viewBox='0 0 24 24'
        strokeWidth={1.5}
        stroke='currentColor'
        className='w-6 h-6'
      >
        <path strokeLinecap='round' strokeLinejoin='round' d='M12 19.5v-15m0 0l-6.75 6.75M12 4.5l6.75 6.75' />
      </svg>
    )
  }
  const IconDown = () => {
    return (
      <svg
        xmlns='http://www.w3.org/2000/svg'
        fill='none'
        viewBox='0 0 24 24'
        strokeWidth={1.5}
        stroke='currentColor'
        className='w-6 h-6'
      >
        <path strokeLinecap='round' strokeLinejoin='round' d='M19.5 13.5L12 21m0 0l-7.5-7.5M12 21V3' />
      </svg>
    )
  }

  return (
    <table className='table-fixed w-full'>
      <thead className='bg-gray-200'>
        <tr className='space'>
          <th className='border px-4 py-2 w-12'>#</th>
          <th className='border px-4 py-2'>Project Name</th>
          <th className='border px-4 py-2'>Hours</th>
          <th className='flex justify-between border px-4 py-2 cursor-pointer ' onClick={orderProjectsByEndDate}>
            DeadLine {up ? <IconUp /> : <IconDown />}
          </th>
        </tr>
      </thead>
      <tbody>
        {globalData.loading && <div>Loading...</div>}
        {globalData.error && <div>Error: {globalData.error}</div>}
        {globalData.projects.map((project) => (
          <Item key={project.id} project={project} />
        ))}
      </tbody>
    </table>
  )
}
