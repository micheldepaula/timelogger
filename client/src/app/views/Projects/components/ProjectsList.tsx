import React, { useState } from 'react'
import Item from './ProjectsListItem'

import { useAppDispatch } from '../../../hooks'
import { sortedProjectsAsc, Project } from '../../../slice/projectSlice'

interface ProjectsListProps {
  projects: Array<Project>
  error?: string
  loading?: string
}
const ProjectsList: React.FC<ProjectsListProps> = ({ projects, error, loading }) => {
  const [up, setUp] = useState<boolean>(true)
  const dispatch = useAppDispatch()

  const orderProjectsByEndDate = () => {
    if (up) {
      dispatch(sortedProjectsAsc(projects))
      setUp(!up)
    } else {
      dispatch({ type: 'projects/sortedProjectsDesc' })
      setUp(!up)
    }
  }

  const IconUp = () => {
    return (
      <svg
        xmlns='http://www.w3.org/2000/svg'
        fill='none'
        viewBox='0 0 24 24'
        strokeWidth={1.5}
        stroke='currentColor'
        className='w-6 h-6'
      >
        <path strokeLinecap='round' strokeLinejoin='round' d='M12 19.5v-15m0 0l-6.75 6.75M12 4.5l6.75 6.75' />
      </svg>
    )
  }
  const IconDown = () => {
    return (
      <svg
        xmlns='http://www.w3.org/2000/svg'
        fill='none'
        viewBox='0 0 24 24'
        strokeWidth={1.5}
        stroke='currentColor'
        className='w-6 h-6'
      >
        <path strokeLinecap='round' strokeLinejoin='round' d='M19.5 13.5L12 21m0 0l-7.5-7.5M12 21V3' />
      </svg>
    )
  }

  return (
    <table className='table-fixed w-full' data-testid='projects-list-component'>
      <thead className='bg-gray-200'>
        <tr className='space'>
          <th className='border px-4 py-2 w-12'>#</th>
          <th className='border px-4 py-2'>Project Name</th>
          <th className='border px-4 py-2'>Hours</th>
          <th className='flex justify-between border px-4 py-2 cursor-pointer ' onClick={orderProjectsByEndDate}>
            DeadLine {up ? <IconUp /> : <IconDown />}
          </th>
        </tr>
      </thead>
      <tbody>
        {loading && <div>Loading...</div>}
        {error && <div>Error: {error}</div>}
        {projects.map((project) => (
          <Item key={project.id} project={project} />
        ))}
      </tbody>
    </table>
  )
}

export default ProjectsList
