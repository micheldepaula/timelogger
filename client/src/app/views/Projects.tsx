import React, { useRef } from 'react'
import { useAppDispatch, useAppSelector } from '../../app/hooks'

import { addProject, projectsSelector } from './../redux/projectSlice'
import Table from '../components/Table'

export default function Projects() {
  const inputNameRef = useRef<HTMLInputElement>(null)
  const inputHoursRef = useRef<HTMLInputElement>(null)
  const inputEndRef = useRef<HTMLInputElement>(null)
  const globalData = useAppSelector(projectsSelector)

  const dispatch = useAppDispatch()

  const handleAddEntry = () => {
    if (inputNameRef.current && inputHoursRef.current && inputEndRef.current) {
      dispatch(
        addProject({
          name: inputNameRef.current.value,
          hours: Number(inputHoursRef.current.value),
          end: new Date(inputEndRef.current.value),
        }),
      )
      inputNameRef.current.value = ''
      inputHoursRef.current.value = ''
      inputEndRef.current.value = ''
    }
  }

  const AddEntry = () => {
    return (
      <div className='w-1/2'>
        <input
          className='border  py-2 px-4'
          type='name'
          ref={inputNameRef}
          placeholder='Project Name'
          aria-label='Project Name'
        />

        <input className='border  py-2 px-4' type='number' ref={inputHoursRef} placeholder='Hours' aria-label='Hours' />

        <input
          className='border  py-2 px-4'
          type='date'
          ref={inputEndRef}
          placeholder='DeadLine'
          aria-label='DeadLine'
        />
        <button
          className='bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded'
          onClick={handleAddEntry}
        >
          Add Entry
        </button>
      </div>
    )
  }

  return (
    <>
      <div className='flex items-center my-6'>
        <AddEntry />
        <div className='w-1/2 flex justify-end'>
          <form>
            <input className='border rounded-full py-2 px-4' type='search' placeholder='Search' aria-label='Search' />
            <button className='bg-blue-500 hover:bg-blue-700 text-white rounded-full py-2 px-4 ml-2' type='submit'>
              Search
            </button>
          </form>
        </div>
      </div>
      {globalData.loading && <div>Loading...</div>}
      {globalData.error && <div>Error: {globalData.error}</div>}
      <Table />
    </>
  )
}
