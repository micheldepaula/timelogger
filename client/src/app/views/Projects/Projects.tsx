import React, { useEffect, useRef } from 'react'
import { useAppDispatch, useAppSelector } from '../../../app/hooks'
import { addProject, getProjects, projectsSelector } from '../../slice/projectSlice'
import ProjectsList from './components/ProjectsList'

const Projects: React.FC = () => {
  const inputNameRef = useRef<HTMLInputElement>(null)
  const inputHoursRef = useRef<HTMLInputElement>(null)
  const inputEndRef = useRef<HTMLInputElement>(null)
  const { projects, loading, error } = useAppSelector(projectsSelector)

  useEffect(() => {
    dispatch(getProjects())
  }, [])

  const dispatch = useAppDispatch()

  const handleAddEntry = () => {
    if (inputNameRef.current && inputHoursRef.current && inputEndRef.current) {
      dispatch(
        addProject({
          name: inputNameRef.current.value,
          hours: Number(inputHoursRef.current.value),
          end: new Date(inputEndRef.current.value),
        }),
      )
      inputNameRef.current.value = ''
      inputHoursRef.current.value = ''
      inputEndRef.current.value = ''
    }
  }

  const AddEntry = () => {
    return (
      <div className='w-1/2'>
        <input
          className='border  py-2 px-4'
          type='name'
          ref={inputNameRef}
          placeholder='Project Name'
          aria-label='Project Name'
          data-testid='project-name-input'
        />

        <input
          className='border  py-2 px-4'
          type='number'
          ref={inputHoursRef}
          placeholder='Hours'
          aria-label='Hours'
          data-testid='hours-input'
        />

        <input
          className='border  py-2 px-4'
          type='date'
          ref={inputEndRef}
          placeholder='DeadLine'
          aria-label='DeadLine'
          data-testid='deadline-input'
        />
        <button
          className='bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded'
          onClick={handleAddEntry}
          data-testid='add-entry-button'
        >
          Add Entry
        </button>
      </div>
    )
  }

  return (
    <div data-testid='projects-view'>
      <div className='flex items-center my-6'>
        <AddEntry />
        <div className='w-1/2 flex justify-end'>
          <form>
            <input
              className='border rounded-full py-2 px-4'
              type='search'
              placeholder='Search'
              aria-label='Search'
              data-testid='search-input'
            />
            <button
              className='bg-blue-500 hover:bg-blue-700 text-white rounded-full py-2 px-4 ml-2'
              type='submit'
              data-testid='search-button'
            >
              Search
            </button>
          </form>
        </div>
      </div>
      {loading && <div data-testid='loading'>Loading...</div>}
      {error && <div data-testid='error'>Error: {error}</div>}
      <ProjectsList projects={projects} />
    </div>
  )
}

export default Projects
