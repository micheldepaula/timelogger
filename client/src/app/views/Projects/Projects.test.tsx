import React from 'react'
import { fireEvent, render } from '@testing-library/react'
import { Wrapper } from './../../components'
import Projects from './Projects'
import * as redux from 'react-redux'

jest.mock('react-redux', () => {
  const actualModule = jest.requireActual('react-redux')
  return {
    __esModule: true,
    ...actualModule,
  }
})

describe('Projects View', () => {
  const rerender = () =>
    render(
      <Wrapper>
        <Projects />
      </Wrapper>,
    )

  it('renders correctly', () => {
    const { asFragment, getByTestId } = rerender()
    expect(getByTestId('projects-view')).toBeDefined()
    expect(asFragment).toMatchSnapshot()
  })

  it('Adding a new entry', () => {
    const useDispatchSpy = jest.spyOn(redux, 'useDispatch')
    const mockDispatchFn = jest.fn()
    useDispatchSpy.mockReturnValue(mockDispatchFn)

    const { asFragment, getByTestId } = rerender()
    expect(getByTestId('projects-view')).toBeDefined()
    expect(getByTestId('project-name-input')).toBeDefined()
    expect(getByTestId('hours-input')).toBeDefined()
    expect(getByTestId('deadline-input')).toBeDefined()
    expect(getByTestId('add-entry-button')).toBeDefined()

    const deadlineInput = getByTestId('deadline-input')
    const hoursInput = getByTestId('hours-input')
    const projectNameInput = getByTestId('project-name-input')
    const addEntryButton = getByTestId('add-entry-button')

    fireEvent.change(projectNameInput, { target: { value: 'ProjectName Test 123' } })
    fireEvent.change(hoursInput, { target: { value: '10' } })
    fireEvent.change(deadlineInput, { target: { value: '29-06-2023' } })

    fireEvent.click(addEntryButton)
    expect(mockDispatchFn).toBeCalled()
    expect(asFragment).toMatchSnapshot()
  })
})
