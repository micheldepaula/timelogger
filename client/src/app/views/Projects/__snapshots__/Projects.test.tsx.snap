// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects View Adding a new entry 1`] = `[Function]`;

exports[`Projects View renders correctly 1`] = `[Function]`;
